pkg/agent/openflow/pipeline.go

```diff
@@ -1363,18 +1411,43 @@ func (f *featurePodConnectivity) l3FwdFlowToGateway() []binding.Flow {
 }
 
 // l3FwdFlowToRemoteViaTun generates the flow to match the packets destined for remote Pods via tunnel.
-func (f *featurePodConnectivity) l3FwdFlowToRemoteViaTun(localGatewayMAC net.HardwareAddr, peerSubnet net.IPNet, tunnelPeer net.IP) binding.Flow {
+func (f *featurePodConnectivity) l3FwdFlowToRemoteViaTun(localGatewayMAC net.HardwareAddr, peerSubnet net.IPNet, tunnelPeer net.IP) []binding.Flow {
 	ipProtocol := getIPProtocol(peerSubnet.IP)
-	return L3ForwardingTable.ofTable.BuildFlow(priorityNormal).
-		Cookie(f.cookieAllocator.Request(f.category).Raw()).
-		MatchProtocol(ipProtocol).
-		MatchDstIPNet(peerSubnet).
-		Action().SetSrcMAC(localGatewayMAC).  // Rewrite src MAC to local gateway MAC.
-		Action().SetDstMAC(GlobalVirtualMAC). // Rewrite dst MAC to virtual MAC.
-		Action().SetTunnelDst(tunnelPeer).    // Flow based tunnel. Set tunnel destination.
-		Action().LoadRegMark(ToTunnelRegMark).
-		Action().GotoTable(L3DecTTLTable.GetID()).
-		Done()
+	buildFlow := func(matcher func(b binding.FlowBuilder) binding.FlowBuilder) binding.Flow {
+		builder := L3ForwardingTable.ofTable.BuildFlow(priorityNormal).
+			Cookie(f.cookieAllocator.Request(f.category).Raw()).
+			MatchProtocol(ipProtocol)
+		builder = matcher(builder)
+		return builder.
+			Action().SetSrcMAC(localGatewayMAC).  // Rewrite src MAC to local gateway MAC.
+			Action().SetDstMAC(GlobalVirtualMAC). // Rewrite dst MAC to virtual MAC.
+			Action().SetTunnelDst(tunnelPeer).    // Flow based tunnel. Set tunnel destination.
+			Action().LoadRegMark(ToTunnelRegMark).
+			Action().GotoTable(L3DecTTLTable.GetID()).
+			Done()
+	}
+	flows := []binding.Flow{
+		// The flow handles packets whose destination IP is in the peer subnet.
+		buildFlow(func(b binding.FlowBuilder) binding.FlowBuilder {
+			return b.MatchDstIPNet(peerSubnet)
+		}),
+	}
+	// If DSR is enabled, packets accessing a DSR Service will not be DNATed on the ingress Node, but EndpointIPField
+	// holds the selected backend Pod IP, we match it and DSRServiceRegMark to send these packets to corresponding Nodes.
```

Comment:
Will be selected Endpoint IP used in remote Node? If it is not used in remote Node, maybe we could add some comments to explain that we only select the EndpointIP to decide the remote Node.

Reply:
How DSR works end to end is explained in the commit message and the comment of the loadBalancerMode variable in proxier.go. l3FwdFlowsToRemoteViaTun is only a portion of the whole, and I just want to make it focus on L3Forwarding. I feel it would repeated to explain it here.

For your question, the EndpointIP selected in ingress Node will not be used in backend Node. See the explaination:

> When a Service's LoadBalancerMode is DSR, the following changes will be applied to the OpenFlow flows and groups:
>
> 1. ClusterGroup will be used by traffic working in DSR mode on ingress Node.
> 2. LocalGroup will be used by traffic working in DSR mode on backend Node.
> 3. Traffic working in DSR mode on ingress Node will be marked and treated specially, e.g. bypassing SNAT.

Reply:
Got that. For `ClusterGroup will be used by traffic working in DSR mode on ingress Node.`, if a remote Endpoint is selected, only the remote Node is decided by removing the suffix of the Endpoint IP, and the packet will be sent to remote Node via tunnel. In remote Node, the Endpoint will be selected finally in a local group. Is that right? If so, how about adding more details to explain that how the remote Endpoint is selected?

Reply:
what details do you mean by "adding more details to explain that how the remote Endpoint is selected?"? LocalGroup will be used on backend Node, so only Endpoints on the remote Node wil be selected if the traffic is sent to it, and the bucket selection is just like all other group selection.

Reply:
I should misunderstood something. I remembered that you have updated the way of bucket selection. My question is that:
- On ingress Node, if a remote Endpoint A is selected in cluster group, then we can get the CIDR of the Node that holds Endpoint A.
- On the Node that holds Endpoint A, Endpoint A* is selected in local group.
- Will Endpoint A and A* be the same?

Reply:
> I remembered that you have updated the way of bucket selection.

No, bucket selection method update is removed due to its impact on session affinity and internal traffic.

> Will Endpoint A and A* be the same?

No, we didn't encode Endpoint A into overlay packet. And it wouldn't require local group to be used on backend Node if they are the same. We DO another selection on backend Node and only pick an Endpoint from local ones.
If you think about it, each Endpoint gets the same chance to be selected regardless of its location. And there is no need to ensure the Endpoint selected on the ingress Node to be the same one selected by backend Node.

I have updated the comment and description to further explain, hope it helps.

Reply:
The updated comment is much helpful to understand that how the remote Endpoint is selected in DSR mode.